### PR TITLE
add `visit_operand` to const prop

### PR DESCRIPTION
**Pull request by lcnr**
r? @oli-obk

### CODE REVIEW

File: src/test/mir-opt/const_prop/array_index/32bit/rustc.main.ConstProp.diff [hunk 1]
```diff
@@ -64,13 +64,25 @@
...
++                                          // + val: Value(Scalar(0x00000004))
++                                          // mir::Constant
++                                          // + span: $DIR/array_index.rs:5:18: 5:33
++                                          // + literal: Const { ty: usize, val: Value(Scalar(0x00000004)) }
++                                          // ty::Const
++                                          // + ty: usize
++                                          // + val: Value(Scalar(0x00000002))
++                                          // mir::Constant
++                                          // + span: $DIR/array_index.rs:5:18: 5:33
++                                          // + literal: Const { ty: usize, val: Value(Scalar(0x00000002)) }
```
Comment (oli-obk) [inline src/test/mir-opt/const_prop/array_index/32bit/rustc.main.ConstProp.diff:85]: oh god these comments are really annoying :laughing:

File: src/test/mir-opt/const_prop/discriminant/32bit/rustc.main.ConstProp.diff [hunk 2]
```diff
@@ -56,7 +56,14 @@
       }
   
       bb2: {
-          switchInt(((_3 as Some).0: bool)) -> [false: bb1, otherwise: bb3]; // scope 0 at $DIR/discriminant.rs:11:26: 11:30
+-         switchInt(((_3 as Some).0: bool)) -> [false: bb1, otherwise: bb3]; // scope 0 at $DIR/discriminant.rs:11:26: 11:30
++         switchInt(const true) -> [false: bb1, otherwise: bb3]; // scope 0 at $DIR/discriminant.rs:11:26: 11:30
```
Comment (oli-obk) [inline src/test/mir-opt/const_prop/discriminant/32bit/rustc.main.ConstProp.diff:60]: awesome! later passes will replace this with a `goto` now

**Comment by oli-obk** [comment]
@bors try @rust-timer build

let's do a perf run so we know whether we can let it get rolled up or not

File: src/librustc_mir/transform/const_prop.rs [hunk 3]
```diff
@@ -1059,31 +1093,10 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
                                 msg,
                             );
                         }
-                    } else {
```
Comment (oli-obk) [inline src/librustc_mir/transform/const_prop.rs:1062]: if you leave this code in, you get both the mir-opt-level 3 downgrade for the advanced features as well as keep the propagation into asserts and switchint. Leave a FIXME for their removal though

File: src/librustc_mir/transform/const_prop.rs [hunk 4]
```diff
@@ -1072,18 +1110,13 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
...
-                        if let ScalarMaybeUninit::Scalar(scalar) =
-                            self.ecx.read_scalar(value).unwrap()
-                        {
-                            *discr = self.operand_from_scalar(scalar, switch_ty, source_info.span);
-                        }
-                    }
-                }
+            TerminatorKind::SwitchInt { ref mut discr, .. } => {
+                // FIXME: This is currently redundant with `visit_operand`, but sadly
+                // always visiting operands currently causes a perf regression, so
```
Comment (xldenis) [inline src/librustc_mir/transform/const_prop.rs:1115]: could be worth mentioning the perf regression is not due to the cost of actually visiting the operands but the changes it causes to MIR code

File: src/test/mir-opt/const_prop/array_index/32bit/rustc.main.ConstProp.diff (hunk 1 above)
Comment (xldenis) [reply src/test/mir-opt/const_prop/array_index/32bit/rustc.main.ConstProp.diff:85]: i really wish there rwas a way to suppress all the comments in MIR dumps

Comment (oli-obk) [reply src/test/mir-opt/const_prop/array_index/32bit/rustc.main.ConstProp.diff:85]: https://github.com/rust-lang/rust/issues/74508

**Comment by oli-obk** [comment]
r=me with more info on the perf regression on that comment

**Comment by lcnr** [comment]
referenced LLVM in the fixme, not sure if you want anything else mentioned.

**Comment by Manishearth** [comment]
Should this be rollup=never?

**Comment by lcnr** [comment]
Not sure, in the end this shouldn't change anything for `mir-opt-level < 2` which is the default so I would be surprised if this has a perf impact

File: src/librustc_mir/transform/const_prop.rs [hunk 5]
```diff
@@ -905,6 +933,16 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
         }
     }
 
+    fn visit_operand(&mut self, operand: &mut Operand<'tcx>, location: Location) {
+        // Only const prop copies and moves on `mir_opt_level=3` as doing so
+        // currently increases compile time.
+        if self.tcx.sess.opts.debugging_opts.mir_opt_level < 3 {
+            self.super_operand(operand, location)
+        } else {
+            self.propagate_operand(operand, location)
```
Comment (RalfJung) [inline src/librustc_mir/transform/const_prop.rs:942]: Why does this not call `super_operand`? Almost all visitors do that; there should be a comment explaining why it does not happen here.

Comment (lcnr) [reply src/librustc_mir/transform/const_prop.rs:942]: aren't we doing this inside of `propagate_operand`? I am not too familiar with this code though?

Should I rewrite it to
```rust
        if self.tcx.sess.opts.debugging_opts.mir_opt_level >= 3 {
            self.propagate_operand(operand, location)
        }
        self.super_operand(operand, location)

Comment (RalfJung) [reply src/librustc_mir/transform/const_prop.rs:942]: > aren't we doing this inside of propagate_operand?

You added that function in this PR, didn't you? I cannot see it doing anything like that, but maybe I am missing the right line.

> Should I rewrite it to

I think that would be better, yes.

Comment (lcnr) [reply src/librustc_mir/transform/const_prop.rs]: done

**Comment by RalfJung** [comment]
LGTM; let's see what @oli-obk says.

**Comment by oli-obk** [comment]
this visits unevaluated constants and potentially evaluates them again, so that change is definitely an improvement

@bors r+

**Pull request merged**